File: python/pig-latin/pig_latin.py
def translate(text):
    """
    Translate text from English to Pig Latin according to the following rules:
    1. If a word begins with a vowel, or starts with "xr" or "yt", add "ay" to the end of the word.
    2. If a word begins with one or more consonants, move those consonants to the end of the word and add "ay".
    3. If a word starts with zero or more consonants followed by "qu", move those consonants (if any) and the "qu" to the end of the word and add "ay".
    4. If a word starts with one or more consonants followed by "y", move the consonants preceding the "y" to the end of the word and add "ay".
    """
    words = text.split()
    translated_words = [translate_word(word) for word in words]
    return " ".join(translated_words)

def translate_word(word):
    # Rule 1: Word begins with a vowel or starts with "xr" or "yt"
    if word[0] in "aeiou" or word.startswith("xr") or word.startswith("yt"):
        return word + "ay"
    
    # Rule 3: Word starts with consonants followed by "qu"
    if "qu" in word:
        qu_index = word.index("qu")
        # Check if "qu" is preceded only by consonants
        if all(letter not in "aeiou" for letter in word[:qu_index]):
            return word[qu_index+2:] + word[:qu_index+2] + "ay"
    
    # Rule 4: Word starts with consonants followed by "y"
    if "y" in word[1:]:  # "y" not at the beginning
        y_index = word.find("y", 1)  # Find "y" starting from the second character
        # Check if "y" is preceded only by consonants
        if all(letter not in "aeiou" for letter in word[:y_index]):
            return word[y_index:] + word[:y_index] + "ay"
    
    # Rule 2: Word begins with one or more consonants
    # Find the index of the first vowel
    for i, letter in enumerate(word):
        if letter in "aeiou" or (letter == "y" and i > 0):
            return word[i:] + word[:i] + "ay"
    
    # If no vowels found, return the original word with "ay"
    return word + "ay"

File: go/pig-latin/pig_latin.go
package piglatin

import (
	"strings"
)

func Sentence(sentence string) string {
	words := strings.Fields(sentence)
	var result []string

	for _, word := range words {
		result = append(result, translateWord(word))
	}

	return strings.Join(result, " ")
}

func translateWord(word string) string {
	// Rule 1: If a word begins with a vowel, or starts with "xr" or "yt", add "ay" to the end
	if isVowel(word[0]) || strings.HasPrefix(word, "xr") || strings.HasPrefix(word, "yt") {
		return word + "ay"
	}

	// Rule 3: Handle words beginning with "qu" or consonants followed by "qu"
	quIndex := strings.Index(word, "qu")
	if quIndex >= 0 && quIndex <= 1 {
		// If "qu" is at the beginning or after one consonant
		return word[quIndex+2:] + word[:quIndex+2] + "ay"
	}

	// Rule 4: Handle words with consonants followed by "y"
	// First check if "y" is not the first letter and is preceded by consonants
	yIndex := strings.IndexRune(word, 'y')
	if yIndex > 0 {
		// Check if all characters before "y" are consonants
		allConsonantsBefore := true
		for i := 0; i < yIndex; i++ {
			if isVowel(word[i]) {
				allConsonantsBefore = false
				break
			}
		}
		
		if allConsonantsBefore {
			return word[yIndex:] + word[:yIndex] + "ay"
		}
	}

	// Rule 2: If a word begins with consonants, move them to the end and add "ay"
	// Find the index of the first vowel (including "y" if not at the beginning)
	firstVowelIndex := len(word)
	for i, c := range word {
		if isVowel(byte(c)) || (c == 'y' && i > 0) {
			firstVowelIndex = i
			break
		}
	}

	if firstVowelIndex < len(word) {
		return word[firstVowelIndex:] + word[:firstVowelIndex] + "ay"
	}

	// Default case (shouldn't reach here with the given rules)
	return word + "ay"
}

func isVowel(c byte) bool {
	return c == 'a' || c == 'e' || c == 'i' || c == 'o' || c == 'u'
}
